lib/spack/spack/installer.py

```diff
@@ -829,7 +829,7 @@ def _prepare_for_install(self, task):
         if restage and task.pkg.stage.managed_by_spack:
             task.pkg.stage.destroy()
 
-        if not partial and self.layout.check_installed(task.pkg.spec) and (
+        if not partial and task.pkg.installed and (
```

Comment:
Short: I agree with this change but I think that instead of this current diff, 

`if not partial and self.layout.check_installed(task.pkg.spec) and ...`

should be replaced with

`if installed_in_db and ...`

And this whole block should move to line 815 (just after the `self._check_db` call.

More details:

Note that `pkg.installed == installed_in_db` (both query the associated DB record `.installed` property), so checking for `not partial` would be redundant (it is only set `True` when `installed_in_db == True`).

This branch assumes there is a database record is present (since it calls `db.update_explicit`). Moreover:

* If `check_installed` is true here, then `os.path.isdir(task.pkg.spec.prefix) == True` above
* `If installed_in_db == False`, then we would have entered the above branch (on line 820)
  * In that case, if we enter this loop, `partial` is False, which means we would have removed the prefix (contradiction)

That means if we enter this loop, `installed_in_db` must be true. Furthermore, I think we would always want to perform this logic when `installed_in_db == True`

Extrapolating: this logic is intended to do the proper bookkeeping for packages that have been "correctly" installed, and its current placement (and the if condition) make that confusing.

Reply:
Okay, I'll go over the code and see if I agree. One thing I noted: `check_installed` is more strict than `pkg.installed` in the sense that it checks for install prefix directory collisions; it verifies the spec.yaml file in the prefix dir is actually the same.

So if you would change the projections to something like `${PACKAGE}-${VERSION}` without the hash, you may have frequent prefix collisions, but not necessarily database entry collisions (?).

Was the intention to guard against prefix dir collisions here?

Reply:
Hm, confusing, we don't hit that code path:

```
root@d8c830a76c06:~# spack -c 'config:install_tree:projections:all:${PACKAGE}' install zlib
==> zlib: Successfully installed zlib-1.2.11-jya4ukqhht537qwxl7cry2sjy64npe5f
[+] /opt/spack/opt/spack/zlib

root@d8c830a76c06:~# spack -c 'config:install_tree:projections:all:${PACKAGE}' install zlib~shared
==> zlib: Successfully installed zlib-1.2.11-dulb6gikvxyooqmq56hkptjbttdxjmz5
[+] /opt/spack/opt/spack/zlib

root@d8c830a76c06:~# spack -c 'config:install_tree:projections:all:${PACKAGE}' find -p
==> 2 installed packages
-- linux-ubuntu16.04-x86_64_v3 / gcc@5.4.0 ----------------------
zlib@1.2.11  /opt/spack/opt/spack/zlib
zlib@1.2.11  /opt/spack/opt/spack/zlib
```